homeassistant/components/thethingsnetwork/__init__.py

```diff
@@ -33,10 +38,57 @@
 
 async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
     """Initialize of The Things Network component."""
-    conf = config[DOMAIN]
-    app_id = conf.get(CONF_APP_ID)
-    access_key = conf.get(CONF_ACCESS_KEY)
 
-    hass.data[DATA_TTN] = {TTN_ACCESS_KEY: access_key, TTN_APP_ID: app_id}
+    if DOMAIN in config:
+        ir.async_create_issue(
+            hass,
+            DOMAIN,
+            "manual_migration",
+            breaks_in_ha_version="2024.6.0",
+            is_fixable=False,
+            severity=ir.IssueSeverity.ERROR,
+            translation_key="manual_migration",
+            translation_placeholders={
+                "domain": DOMAIN,
+                "v2_v3_migration_url": "https://www.thethingsnetwork.org/forum/c/v2-to-v3-upgrade/102",
+                "v2_deprecation_url": "https://www.thethingsnetwork.org/forum/t/the-things-network-v2-is-permanently-shutting-down-completed/50710",
+            },
+        )
+
+    return True
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Establish connection with The Things Network."""
+
+    _LOGGER.debug(
+        "Set up %s at %s",
+        entry.data[CONF_API_KEY],
+        entry.data.get(CONF_HOST, TTN_API_HOST),
+    )
+
+    coordinator = TTNCoordinator(hass, entry)
+
+    await coordinator.async_config_entry_first_refresh()
+
+    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = coordinator
```

Comment:
After we merge this, please checkout the ConfigEntry typing, there has been a blog post about it recently. It's a good thing for a follow-up pr